File: 2021/day03_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any


def solution(content: list[str]) -> Any:
    parsed = [parse(line) for line in content if line]
    acc = count_all(parsed)
    return to_int(most_common_value(acc)) * to_int(least_common_value(acc))


def to_int(binary: list[int]):
    return sum(bit * 2**index for index, bit in enumerate(reversed(binary)))


def most_common_value(count: list[tuple[int, int]]):
    return [0 if c0 > c1 else 1 for c0, c1 in count]


def least_common_value(count: list[tuple[int, int]]):
    return [0 if c0 < c1 else 1 for c0, c1 in count]


def init(length: int):
    return [(0, 0)] * length


def count_all(content: list[list[int]]):
    acc = init(len(content[0]))
    for line in content:
        acc = count_line(acc, line)
    return acc


def count_line(acc: list[tuple[int, int]], line: list[int]):
    return [
        (c0 + (1 if bit == 0 else 0), c1 + (1 if bit == 1 else 0))
        for ((c0, c1), bit) in zip(acc, line)
    ]


def parse(line: str):
    return [int(x) for x in line.strip()]


def main() -> None:
    with open("day03.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2021/day03-1.ts
type Count = { c0: number; c1: number };

function solution(content: string) {
  const count = countAll(
    content.split("\n").map((line) => line.trim().split("").map(Number))
  );
  return toInt(mostCommonValue(count)) * toInt(leastCommonValue(count));
}

function toInt(binary: number[]) {
  return binary
    .reverse()
    .reduce((acc, cur, index) => acc + cur * Math.pow(2, index), 0);
}

function mostCommonValue(count: Count[]) {
  return count.map(({ c0, c1 }) => (c0 > c1 ? 0 : 1));
}
function leastCommonValue(count: Count[]) {
  return count.map(({ c0, c1 }) => (c0 < c1 ? 0 : 1));
}

function countAll(content: number[][]) {
  return content.reduce(
    (acc, cur) =>
      acc.map(({ c0, c1 }, index) => ({
        c0: c0 + (cur[index] === 0 ? 1 : 0),
        c1: c1 + (cur[index] === 1 ? 1 : 0),
      })),
    Array.from({ length: content[0].length }).map(() => ({
      c0: 0,
      c1: 0,
    })) as Count[]
  );
}

Deno.readTextFile("day03.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
